Doc/library/importlib.rst

```diff
@@ -1282,68 +1281,81 @@ find and load modules.
-   the module's ``__path__`` at runtime, and this will not be automatically
-   reflected in ``__spec__.submodule_search_locations``.
+   attribute available directly on the module object, e.g.
+   ``module.__spec__.origin == module.__file__``.  Note, however, that while
+   the *values* are usually equivalent, they can differ since there is no
+   synchronization between the two objects.  For example, it is possible
+   to update the module's :attr:`__file__` at runtime and this will not be
+   automatically reflected in the module's :attr:`__spec__.origin`, and
+   vice versa.
 
    .. versionadded:: 3.4
 
    .. attribute:: name
 
-   (``__name__``)
+   (:attr:`__name__`)
 
-   A string for the fully-qualified name of the module.
+   The module's fully-qualified name.  Finders should always set this to a
+   non-empty (identifier) string.
 
    .. attribute:: loader
 
-   (``__loader__``)
+   (:attr:`__loader__`)
 
-   The loader to use for loading.  For namespace packages this should be
-   set to ``None``.
+   The :class:`Loader <importlib.abc.Loader>` that should be used when
+   loading the module.  Finders should always set this.
 
    .. attribute:: origin
 
-   (``__file__``)
+   (:attr:`__file__`)
 
-   Name of the place from which the module is loaded, e.g. "builtin" for
-   built-in modules and the filename for modules loaded from source.
-   Normally "origin" should be set, but it may be ``None`` (the default)
-   which indicates it is unspecified (e.g. for namespace packages).
+   The name of where the module should be loaded from (e.g. ``'builtin'``
+   for built-in modules and the path for modules loaded from source).
+   Finders should normally set this attribute, but it may be ``None`` (the
+   default) which indicates it is unspecified (like for namespace packages).
 
    .. attribute:: submodule_search_locations
 
-   (``__path__``)
+   (:attr:`__path__`)
 
-   List of strings for where to find submodules, if a package (``None``
-   otherwise).
+   For packages, the Finder must set this to the iterable specifying the
+   locations where submodules may be found for the package, even if
+   it's just an empty list.
+   For non-package modules, this should be set to ``None``.
+   For namespace packages (where :attr:`origin` is ``None``), this
+   is later set to a special object by the import system.
 
    .. attribute:: loader_state
 
-   Container of extra module-specific data for use during loading (or
-   ``None``).
+   The Finder may set this to an object containing additional,
+   module-specific data to use when loading the module.  Otherwise it
+   should be set to ``None``.
 
    .. attribute:: cached
 
-   (``__cached__``)
+   (:attr:`__cached__`)
 
-   String for where the compiled module should be stored (or ``None``).
+   Where a compiled version of the code should be stored.  For modules
+   that do not need compiled code stored, ``cached`` should be ``None``.
```

Comment:
Interesting!  This looks like another inconsistency:

```
Python 3.9.0+ (heads/3.9:6e842bcdf8, Oct 21 2020, 14:53:22) 
[Clang 12.0.0 (clang-1200.0.32.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.__cached__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'sys' has no attribute '__cached__'
>>> sys.__spec__.cached
```

Reply:
This is not an inconsistency to me, this is consistent with PEP 451 rules. Here are the 3 real inconsistencies that I had found: https://bugs.python.org/issue37519